### DISCUSSION TITLE
Issues with lets-plot Visualization in Molten and MoltenOpenInBrowser

### DISCUSSION THREAD

**Discussion by bassamsdata** [benlubas/molten-nvim#140] [Q&A] [answered]
Hello,

I've been using the [lets-plot](https://lets-plot.org/) library for visualization and encountered two issues:

1. When using Molten with image.nvim, the plot doesn't display even with the option `LetsPlot.setup_html(no_js=True)`. However, MoltenOpenInBrowser shows the plot.
output: 
```lua
Out[4]: ✓ Done
<lets_plot.plot.core.PlotSpec at 0x10b11d890>
```

2. Without setting the option (needing JS visibility), MoltenOpenInBrowser doesn't display any plots in the browser. This issue also occurs with plotly.

#### to reproduce:
this [jupyter file](https://github.com/bassamsdata/Sila/blob/b14faa2b2110c137cfb08c4be24fba47274d2cf8/testing_molten.ipynb) I created for the issue:

My questions: Is this normal behavior? If yes, I suggest an idea: Could we consider displaying all plots in a browser window without the need for Async (which I know it might be hard)? This would be updated as e.g when saving a notebook. I envision keeping the browser open with a history of all plots (similar to what I've seen in #138). This idea is inspired by a practice in R programming that I use daily with [httpgd](https://github.com/nx10/httpgd).

Any insights or considerations would be much appreciated!

Thank you

**Comment by benlubas** [comment]
The plot not displaying as an image isn't surprising, it seems like the program produces an HTML output and not an image output. In order to render the plots in molten as an image we would have to do something special like we're doing for plotly and render the image in molten before displaying it.

I'm curious about needing to set `no_js=True`, it seems to me like adding JS shouldn't effect how we write the HTML to a file and the tell the browser to open it.

I wonder if it has a different Mimetype when it has JS actually. Could you turn on debugging by setting the `vim.g.molten_show_mimetype_debug = true` and tell me what the mimitypes are?

> I envision keeping the browser open with a history of all plots

This would be cool, but i think it would require some type of local webserver to work the way you want. Otherwise you would be stuck refreshing the page yourself.

**Reply by bassamsdata** [reply]
thank you for your answer. 

> I wonder if it has a different Mimetype when it has JS actually. Could you turn on debugging by setting the vim.g.molten_show_mimetype_debug = true and tell me what the mimitypes are?

actually, it has the same mimitypes for the this plot with and without the option `no_js=True`
```lua
Out[25]: ✓ Done
[DEBUG] Received mimetypes: ['text/plain', 'text/html']
<lets_plot.plot.core.PlotSpec at 0x11a3e2290>
```
<img width="711" alt="Screenshot 2024-02-03 at 8 06 26 PM" src="https://github.com/benlubas/molten-nvim/assets/105807570/36c875cb-3d3e-4a77-aeaa-d930a867961b">



yet, when I don't put the option to not use js, this mimitype appears: (when using the option, the output will be `Done` only.)
```lua
Out[26]: ✓ Done
[DEBUG] Received mimetypes: ['text/html']
<No usable MIMEtype! Received mimetypes ['text/html']>
```
<img width="576" alt="Screenshot 2024-02-03 at 8 10 17 PM" src="https://github.com/benlubas/molten-nvim/assets/105807570/c1504e94-c6ab-4818-8a58-b3af28168a90">


> This would be cool, but i think it would require some type of local webserver to work the way you want. Otherwise you would be stuck refreshing the page yourself.

tbh, I wouldn't mind any solution, as long as I can see the plots with no issues. (and I think saving plots and enlarge them would be easier ). btw, thank you for other plugin that save the plots.

**Reply by benlubas** [reply]
If I had to guess, this is something similar to the problem originally described [here](https://github.com/benlubas/molten-nvim/issues/142), where when you enable javascript there's some type of library that's supposed to be auto injected that molten doesn't pick up on.

If you open the browser on an output that has JS enabled and open the console do you see similar errors to when I sent in that issue?

**Reply by bassamsdata** [reply]
I agree that's probably the same issue.
I searched a bit about it and thought it might be related to a specific browser or plugin, but the issue persisted on many browsers.
I have this console error in the browser:
<img width="776" alt="Screenshot 2024-02-03 at 11 38 48 PM" src="https://github.com/benlubas/molten-nvim/assets/105807570/90a29e09-c193-48d4-86de-14963d4c3a48">

and this the text of the file mentioned in the error:
<details>
  <summary>Code</summary>

  ```html
   <div id="g2Z4fh"></div>
   <script type="text/javascript" data-lets-plot-script="plot">
       (function() {
           var plotSpec={
"data":{
"Name":["John","Anna","Peter","Linda"],
"Salary":[50000.0,55000.0,60000.0,65000.0]
},
"mapping":{
"x":"Name",
"y":"Salary"
},
"data_meta":{
},
"kind":"plot",
"scales":[],
"layers":[{
"geom":"bar",
"stat":"identity",
"mapping":{
},
"data_meta":{
},
"data":{
}
}],
"metainfo_list":[]
};
           var plotContainer = document.getElementById("g2Z4fh");
           window.letsPlotCall(function() {{
               LetsPlot.buildPlotFromProcessedSpecs(plotSpec, -1, -1, plotContainer);
           }});
       })();    
   </script>
  ```
</details>

**Reply by benlubas** [reply]
Yeah, seems like some type of missing injection. I'm really not sure what to do about that. If you keep poking around and come up with a way to get things working I'd be happy to add support.

I'm even less sure about how to handle this case honestly. The linked issue returned its own mimetype, which makes it easy to detect. This one just returns plan old html so I'm not even sure how to detect it, let along which js to inject and how to inject it properly...

**Reply by bassamsdata** [reply]
No problem at all, and I appreciate your effort in looking into it. 

Even though I'm not sure what's happening either, I'll continue using the option for no JS. 

I'll do some searching, and if I come across a solution, I'll post it here. I'll leave the question open in case someone else encounters it and has a workaround.Thank you!

**Reply by bassamsdata** [reply, marked as answer]
I think I found a solution for this from the lets-plot library that I didn't know before.
setting this will always open the plot in full html/javascript mode.
`LetsPlot.setup_show_ext()`
